scipy/special/tests/test_basic.py

```diff
@@ -1319,6 +1319,31 @@ def test_comb(self):
         expected = 100891344545564193334812497256
         assert_equal(special.comb(100, 50, exact=True), expected)
 
+        with pytest.warns(
+            DeprecationWarning,
+            match=r"Non-integer arguments are currently being cast to",
+        ):
+            assert_equal(
+                special.comb(4.5, 3, exact=True, legacy=True),
+                special.comb(4, 3, exact=True),
+            )
+        with pytest.warns(
+            DeprecationWarning,
+            match=r"Non-integer arguments are currently being cast to",
+        ):
+            assert_equal(
+                special.comb(4, 3.5, exact=True, legacy=True),
+                special.comb(4, 3, exact=True),
+            )
+        assert_equal(
+            special.comb(4.5, 3, exact=True, legacy=False),
+            special.comb(4.5, 3),
+        )
+        assert_equal(
+            special.comb(4, 3.5, exact=True, legacy=False),
+            special.comb(4, 3.5),
+        )
```

Comment:
First off, sorry for the invasive proposal, but this is a textbook candidate of parametrization IMO. We need to add tests for `repartition=True` in any case, and that becomes just an extra line with parametization. :)
```suggestion
    @pytest.mark.parametrize("repartition", [True, False])
    @pytest.mark.parametrize("legacy", [True, False])
    @pytest.mark.parametrize("k", [3.5, 3])
    @pytest.mark.parametrize("N", [4.5, 4])
    def test_comb_legacy(N, k, legacy, repartition)
        # test is only relevant for exact=True
        if legacy and (N != int(N) or k != int(k)):
            with pytest.warns(
                DeprecationWarning,
                match=r"Non-integer arguments are currently being cast to",
            ):
                result = special.comb(N, k, exact=True, legacy=legacy,
                                      repartition=repartition)
        else: 
            result = special.comb(N, k, exact=True, legacy=legacy,
                                  repartition=repartition)

        if legacy:
            # for exact=True and legacy=True, cast input arguments, else don't
            N, k = int(N), int(k)
        # expected result is the same as with exact=False
        expected = special.comb(N, k, legacy=legacy, repartition=repartition)
        assert_equal(result, expected)
```

Reply:
I agree with this.

Reply:
Thanks for this suggestion I am far more familiar with unittest than pytest so this is useful to know!

Reply:
This test required a bit of tweaking. Take the case of `N=4.5`, `k=3.5`, `legacy=True` and `repetition=True`. `result` is equal to `comb(4.5, 3.5, exact=True, legacy=True, repetition=True)` this in turn hits line `2225` where `comb` called again as `comb(7, 3.5, exact=True, legacy=True)`. This gives **35**. `expected` is equal to `comb(4, 3, legacy=True, repetition=True)` which when it hits line `2225` calls `comb(6, 3, exact=False, legacy=True)` which is **20**. I fixed this with the following tweak:
```
@pytest.mark.parametrize("repetition", [True, False])
    @pytest.mark.parametrize("legacy", [True, False])
    @pytest.mark.parametrize("k", [3.5, 3])
    @pytest.mark.parametrize("N", [4.5, 4])
    def test_comb_legacy(self, N, k, legacy, repetition):
        # test is only relevant for exact=True
        if legacy and (N != int(N) or k != int(k)):
            with pytest.warns(
                DeprecationWarning,
                match=r"Non-integer arguments are currently being cast to",
            ):
                result = special.comb(N, k, exact=True, legacy=legacy,
                                      repetition=repetition)
        else:
            result = special.comb(N, k, exact=True, legacy=legacy,
                                  repetition=repetition)
        if legacy:
            # for exact=True and legacy=True, cast input arguments, else don't
            if repetition:
                N, k = int(N + k - 1), int(k)
                repetition = False
            else:
                N, k = int(N), int(k)
        # expected result is the same as with exact=False
        expected = special.comb(N, k, legacy=legacy, repetition=repetition)
        assert_equal(result, expected)
```
This is effectively does what repetition does but avoids the round error cause by casting to `ints` before doing `N + k -1`.

Reply:
Good find & resolution, thanks!